pkg/agent/flexvolume/server.go

```diff
@@ -0,0 +1,155 @@
+
+	// remove unix socket if it existed previously
+	if _, err := os.Stat(unixSocketFile); !os.IsNotExist(err) {
+		logger.Info("Deleting unix domain socket file.")
+		os.Remove(unixSocketFile)
+	}
+
+	s.listener, err = net.Listen("unix", unixSocketFile)
+	if err != nil {
+		return fmt.Errorf("unable to listen at %s: %v", unixSocketFile, err)
+	}
+	go rpc.Accept(s.listener)
+
+	if err := os.Chmod(unixSocketFile, 0770); err != nil {
+		return fmt.Errorf("unable to set file permission to unix socket %s: %v", unixSocketFile, err)
+	}
+	logger.Info("Listening on unix socket for Kubernetes volume attach commands.")
+	return nil
+}
+
+// Stop the unix domain socket server and deletes the socket file
+func (s *FlexvolumeServer) Stop() {
+	if s.listener != nil {
+		logger.Info("Stopping unix socket rpc server.")
+		if err := s.listener.Close(); err != nil {
+			logger.Errorf("Failed to stop unix socket rpc server: %+v", err)
+		}
+	}
+	// closing the listener should remove the unix socket file. But lets try it remove it just in case.
+	unixSocketFile := path.Join(flexVolumeDriverDir, path.Join(UnixSocketName)) // /flextmnt/rook.io~rook/.rook.sock
+	if _, err := os.Stat(unixSocketFile); !os.IsNotExist(err) {
+		logger.Info("Deleting unix domain socket file.")
+		os.Remove(unixSocketFile)
+	}
+
+}
+
+func configureFlexVolume(driverFileName, driverDir string) error {
+	// copying flex volume
+	if _, err := os.Stat(driverDir); os.IsNotExist(err) {
+		os.Mkdir(driverDir, 0755)
+	}
+	srcFile := path.Join("/", driverFileName)                          // /flexvolume
+	destFile := path.Join(driverDir, "."+FlexvolumeDriver)             // /flextmnt/rook.io~rook/.rook
+	finalDestFile := path.Join(driverDir, path.Join(FlexvolumeDriver)) // /flextmnt/rook.io~rook/rook
+	err := copyFile(srcFile, destFile)
+	if err != nil {
+		return fmt.Errorf("unable to copy flexvolume from %s to %s: %+v", srcFile, destFile, err)
+	}
+
+	// renaming flex volume. Rename is an atomic execution while copying is not.
+	if _, err := os.Stat(finalDestFile); !os.IsNotExist(err) {
+		// Delete old plugin if it exists
+		err = os.Remove(finalDestFile)
+		if err != nil {
+			logger.Warningf("Could not delete old Rook Flexvolume driver at %s: %v", finalDestFile, err)
+		}
+
+	}
+	return os.Rename(destFile, finalDestFile)
+}
+
+func copyFile(src, dest string) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return fmt.Errorf("error opening source file %s: %v", src, err)
+	}
+	defer srcFile.Close()
+
+	destFile, err := os.OpenFile(dest, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0755) // creates if file doesn't exist
+	if err != nil {
+		return fmt.Errorf("error creating destination file %s: %v", dest, err)
+	}
+	defer destFile.Close()
+
+	_, err = io.Copy(destFile, srcFile)
+	if err != nil {
+		return fmt.Errorf("error copying file from %s to %s: %v", src, dest, err)
+	}
+	defer destFile.Close()
```

Comment:
this is the second `defer destFile.Close()` call in this function.  only 1 should be needed, right?

Reply:
you are right. good catch